docs/architecture/deployment-execution/state.md

```diff
@@ -63,6 +63,227 @@ were to do snapshot updates in parallel, it would be possible that we overwrite
 a snapshot with more information, with one that has been generated earlier, that
 doesn't include the latest updates yet.
 
+(snapshot-journaling)=
+### Snapshot Journaling
+
+To avoid the problem of not being able to send snapshot updates in parallel, and
+to save on bandwidth costs, we are implementing a journaling approach. Instead of
+sending the full snapshot we can create a journal entry for each step, and replay
+them in sequence on top of the current snapshot, to create a new valid snapshot.
+
+This can be done due to the fact that that the engine and the snapshot manager can
+start with the same base snapshot, on top of which the entries are applied. And
+because the engine never starts a new update before any of the dependents have
+completed, it is always safe to apply all current journal entries in order, to
+reconstruct the snapshot. In particular this is different from the snapshotting
+implementation in that we can't rely on the engine setting fields in the snapshot
+anymore, but need to encode all of this information in the journal entries.
+
+In the first pass of the implementation, we still send the whole snapshot to the
+backend. In later stages of the implementation, backends that can handle it, can
+receive the journal entries directly, and rebuild the snapshot on the backend.
+
+To make sure we always operate on the same snapshot the backend has access to,
+and to be able to confidently test the implementation against the lifecycletests,
+the journaler only has access to a copy of the snapshot, where all resources are
+deep copied. This way we make sure that we don't make use of snapshot entries
+being modified by the engine.
+
+The snapshot may be updated once at the beginning, for provider migrations. In
+that case we do a "write" operation update the snapshot. It is not valid to do
+this after any journal entries have been created, because resources might now be
+in different places in the resource list.
+
+#### Journal entry details
+
+There's various types of Journal entries, all with slightly different semantics.
+This section describes them. All journal entries are associated with increasing
+IDs. These IDs are used for the following:
```

Comment:
Do entry IDs start at 0?

Reply:
They start at 1, so we avoid the same problem as we have for `DeleteOld`.  I'll add a line to the doc.